src/group_theory/divisible.lean

```diff
@@ -0,0 +1,299 @@
+* `add_monoid.smul_surj_of_divisible_by` : the constructive definition of divisiblity implies
+  the condition that `n • x = a` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `add_monoid.divisible_by_prod` : `A × B` is divisible for any two divisible additive monoids.
+* `add_monoid.divisible_by_pi` : any product of divisble additive monoids is divisible.
+* `add_group.divisible_by_int_of_divisible_by_nat` : for additive groups, int divisiblity is implied
+  by nat divisiblity.
+* `add_group.divisible_by_nat_of_divisible_by_int` : for additive groups, nat divisiblity is implied
+  by int divisiblity.
+* `add_comm_group.divisible_of_smul_top_eq_top` : the constructive definition of divisiblity is
+  implied by the condition that `n • A = A` for all `n ≠ 0`.
+* `add_comm_group.smul_top_eq_top_of_divisible` : the constructive definition of divisiblity implies
+  the condition that `n • A = A` for all `n ≠ 0`.
+* `add_comm_group.divisible_of_char_zero` : any field of characteristic zero is divisible.
+* `add_comm_group.divisible_by_quotient` : 1uotient group of divisible group is divisible.
+* `add_comm_group.divisible_by_of_surj` : if `A` is divisible and `A →+ B` is surjective, then `B`
+  is divisible.
+
+and their multiplicative counterparts:
+* `monoid.rootable_by_of_pow_surj` : the constructive definition of rootablity is implied by the
+  condition that `xⁿ = y` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `monoid.pow_surj_of_rootable_by` : the constructive definition of rootablity implies the
+  condition that `xⁿ = y` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `monoid.rootable_by_prod` : any product of two rootable monoids is rootable.
+* `monoid.rootable_by_pi` : any product of rootable monoids is rootable.
+* `group.rootable_by_int_of_rootable_by_nat` : in groups, int rootablity is implied by nat
+  rootablity.
+* `group.rootable_by_nat_of_rootable_by_int` : in groups, nat rootablity is implied by int
+  rootablity.
+* `comm_group.rootable_by_quotient` : quotient group of rootable group is rootable.
+* `comm_group.rootable_by_of_surj` : if `A` is rootable and `A →* B` is surjective, then `B` is
+  rootable.
+
+TODO: Show that divisibility implies injectivity in the category of `AddCommGroup`.
+-/
+
+open_locale pointwise
+
+namespace add_monoid
+
+variables (A α : Type*) [add_monoid A] [has_scalar α A] [has_zero α]
+
+/--
+An `add_monoid A` is `α`-divisible iff `n • x = a` has a solution for all `n ≠ 0 ∈ α` and `a ∈ A`.
+Here we adpot a constructive approach where we ask an explicit `div : A → α → A` function such that
+* `div a 0 = 0` for all `a ∈ A`
+* `n • div a n = a` for all `n ≠ 0 ∈ α` and `a ∈ A`.
+-/
+class divisible_by :=
+(div : A → α → A)
+(div_zero : ∀ a, div a 0 = 0)
+(div_cancel : ∀ {n : α} (a : A), n ≠ 0 → n • (div a n) = a)
+
+section pi
+
+variables {ι β : Type*} (B : ι → Type*) [Π (i : ι), has_scalar β (B i)]
+
+instance has_scalar_pi : has_scalar β (Π i, B i) :=
+{ smul := λ n x i, n • (x i) }
```

Comment:
This already exists, `import group_theory.group_action.pi` if it's not already imported (it may be in a different file)